### PR TITLE
Replaced Qt dependency with Boost Meta-state machine

### PR DESCRIPTION
**Pull request by joshuaplusone**
@shaun-edwards still needs more testing, but the code is in place.

### CODE REVIEW

File: packml_sm/CMakeLists.txt [hunk 1]
```diff
@@ -1,65 +1,51 @@
...
-
-if("${qt_gui_cpp_USE_QT_MAJOR_VERSION} " STREQUAL "5 ")
-  qt5_wrap_cpp(packml_sm_MOCS ${packml_sm_HDRS})
-  #qt5_wrap_ui(packml_sm__UIS_H ${packml_sm_UIS})
-else()
-  qt4_wrap_cpp(packml_sm_MOCS ${packml_sm_HDRS})
-  #qt4_wrap_ui(packml_sm__UIS_H ${packml_sm_UIS})
-endif()
-
+add_definitions(-DBOOST_MPL_CFG_NO_PREPROCESSED_HEADERS -DBOOST_MPL_LIMIT_VECTOR_SIZE=60 -DBOOST_MPL_LIMIT_MAP_SIZE=60 -DFUSION_MAX_VECTOR_SIZE=50)
```
Comment (shaun-edwards) [inline packml_sm/CMakeLists.txt]: So I looked these up and it wasn't obvious why they needed to be changed from their defaults.  Am I missing something?

File: packml_sm/CMakeLists.txt [hunk 2]
```diff
@@ -1,65 +1,51 @@
...
-#add_dependencies(${PROJECT_NAME} ${catkin_EXPORTED_TARGETS})
+add_library(${PROJECT_NAME} ${packml_sm_SRCS})
 target_link_libraries(${PROJECT_NAME} ${catkin_LIBRARIES})
-if("${qt_gui_cpp_USE_QT_MAJOR_VERSION} " STREQUAL "5 ")
-  target_link_libraries(${PROJECT_NAME} Qt5::Widgets)
-else()
-  target_link_libraries(${PROJECT_NAME} ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY})
-endif()
-
+target_compile_definitions(${PROJECT_NAME} PUBLIC -DBOOST_MPL_CFG_NO_PREPROCESSED_HEADERS PUBLIC -DBOOST_MPL_LIMIT_VECTOR_SIZE=60 PUBLIC -DBOOST_MPL_LIMIT_MAP_SIZE=60 PUBLIC -DFUSION_MAX_VECTOR_SIZE=50)
```
Comment (shaun-edwards) [inline packml_sm/CMakeLists.txt:29]: More education for me...what does the `PUBLIC` keyword do?

File: packml_sm/cmake/packml_sm-extras.cmake [hunk 3]
```diff
@@ -0,0 +1 @@
+set(CMAKE_CXX_FLAGS "-DBOOST_MPL_CFG_NO_PREPROCESSED_HEADERS -DBOOST_MPL_LIMIT_VECTOR_SIZE=60 -DBOOST_MPL_LIMIT_MAP_SIZE=60 -DFUSION_MAX_VECTOR_SIZE=50 ${CMAKE_CXX_FLAGS}")
```
Comment (shaun-edwards) [inline packml_sm/cmake/packml_sm-extras.cmake:1]: Three times?

File: packml_sm/include/packml_sm/boost/packml_events.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <boost/mpl/map/map50.hpp>
+
+namespace packml_sm
+{
+struct clear_event
```
Comment (shaun-edwards) [inline packml_sm/include/packml_sm/boost/packml_events.h:24]: What are these empty structs used for?

File: packml_sm/include/packml_sm/boost/packml_events.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Software License Agreement (Apache License)
+ *
+ * Copyright (c) 2018 Joshua Curtis
```
Comment (shaun-edwards) [inline packml_sm/include/packml_sm/boost/packml_events.h]: Can you show Plus One Robotics some love in the copyright?  

I wrote packml before Plus One existed, hence the lack of it in my headers.  (BTW, feel free to add (i.e keep multiple) a copyright to the other headers that you changed/refactored).

Comment (shaun-edwards) [reply packml_sm/include/packml_sm/boost/packml_events.h]: Multiple headers...

File: packml_sm/include/packml_sm/make_unique.h [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+// to have syntax compatibility with C++14.
+//
+// I perform compiler version checking for MSVC, gcc, and clang to ensure that
+// we don't add make_unique if it is already there (instead, we include
+// <memory> to get the compiler-provided one). You can override the compiler
+// version checking by defining the symbol COMPILER_SUPPORTS_MAKE_UNIQUE.
+//
+//
+// ===============================================================================
+// This file is released into the public domain. See LICENCE for more information.
```
Comment (shaun-edwards) [inline packml_sm/include/packml_sm/make_unique.h:48]: Is this license anywhere?

**Review by shaun-edwards — CHANGES_REQUESTED** [review]
Mostly questions/clarifications for myself.  One general thought is that we really should document this better if the Boost MSM is the structure we plan to stick with (I believe this is true).

File: packml_sm/CMakeLists.txt (hunk 2 above)
Comment (joshuaplusone) [reply packml_sm/CMakeLists.txt:29]: Allows depending targets to also use the definitions.

File: packml_sm/cmake/packml_sm-extras.cmake (hunk 3 above)
Comment (joshuaplusone) [reply packml_sm/cmake/packml_sm-extras.cmake:1]: I changed it to 2X.  Was the only way I could get `catkin build` to work as well as catkin_make.

File: packml_sm/include/packml_sm/boost/packml_events.h (hunk 4 above)
Comment (joshuaplusone) [reply packml_sm/include/packml_sm/boost/packml_events.h:24]: Boost is strongly type-based as most of their implementation is template driven.

File: packml_sm/include/packml_sm/boost/packml_events.h (hunk 5 above)
Comment (joshuaplusone) [reply packml_sm/include/packml_sm/boost/packml_events.h]: Done

File: packml_sm/include/packml_sm/make_unique.h (hunk 6 above)
Comment (joshuaplusone) [reply packml_sm/include/packml_sm/make_unique.h:48]: I appended the license to the top of the file.  https://github.com/bstreiff/cppbits/blob/master/LICENSE

**Comment by shaun-edwards** [comment]
FYI, there is a package for generic logging, found here: http://wiki.ros.org/console_bridge.

It achieves the generic logging with/without ROS and might be a bit more full featured (maybe).  Anyway, it's simple to use.

Here's an example of it's use: https://github.com/ros-industrial-consortium/descartes/pull/227

**Comment by shaun-edwards** [comment]
Did we also lose the timing stats for each state?  See PR (outdated) here: https://github.com/ros-industrial-consortium/packml/pull/32/files

**Comment by joshuaplusone** [comment]
Yes, unintentional though.  I will fix it.
________________________________

**Comment by shaun-edwards** [comment]
@joshuaplusone, thanks for the contribution.

**Pull request merged**